### PR TITLE
New addon: Compact menu bar user dropdown

### LINKED ISSUE #4648: Adjustments to some menu bar items to save space
For those with a small screen, the menu bar can get crowded, especially with addons that add to it. I have some ideas for how we could save space by providing options and updates that shorten or remove some text like so:

Scratch's editor menu bar but modified

### What changed in this image:
- (From an addon) **"Start Recording" changed to "Record"**
- **"Tutorials" text hidden, leaving just the icon**
- **"See Project Page" changed to "Project Page"**
- **Username hidden**

Other ideas to save space:
- (From an addon) **"blocks" in "42 blocks" replaced with an icon**

When I shrunk the window down, nothing was forced to compress or hide to fit thanks to the shorter menu bar.

These are just ideas, and not all of these have to be implemented, or there may be another change we can make, but one or more of these changes in the form of a new addon for Scratch's menu bar items and updates to existing addons for anything added by them would be nice for those with a smaller screen.

Kind of related to #2110

### PR DESCRIPTION
**Pull request by iqnite**
New version at #6297
Resolves #4648
Related to #4018

### Changes

image

With this new addon, users can hide the username from the editor menu bar, leaving only the profile picture. The name is added to the dropdown menu, like in the website implementation.

### Reason for changes

It saves space, which is suitable for small screens, and there's the same feature for the website.

### Tests

_Tested on Vivaldi 5.7.2921._
~~The username hides appropriately from the menu bar, but it is shown in the account menu only one time, when clicking outside and re-opening it the username disappears. Any ideas on how I could fix this so that it's always visible in the menu (like in the website implementation)?~~

~~**Update:** They're now always shown (thanks @CST1229 ), but they are visible on every menu item (so there are 4 displays). Fix?~~

**Update:** All issues mentioned above were addressed thanks to @CST1229.

### CODE REVIEW

File: addons/compact-menu-bar/userscript.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+export default async function ({ addon }) {
+  let span;
+  let li;
+  if (await addon.auth.fetchIsLoggedIn()) {
+    while (true) {
+      // Remove username from bar
+      const username = await addon.auth.fetchUsername();
+      const dropdown = await addon.tab.waitForElement(".account-nav_profile-name_2oRiV");
+      span = dropdown;
+      span?.remove();
```
Comment (Samq64) [inline addons/compact-menu-bar/userscript.js:10]: > The username hides appropriately from the menu bar, but it is shown in the account menu only one time, when clicking outside and re-opening it the username disappears. Any ideas on how I could fix this?

The problem is the script is waiting for this element to be recreated before moving on to the next one which only happens when re-entering the editor.

We prefer using CSS over JavaScript when possible, so the best solution is to move this to the userstyle with `display: none;` instead.

File: addons/compact-menu-bar/userscript.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+export default async function ({ addon }) {
+  let span;
+  let li;
+  if (await addon.auth.fetchIsLoggedIn()) {
+    while (true) {
+      // Remove username from bar
+      const username = await addon.auth.fetchUsername();
```
Comment (Samq64) [inline addons/compact-menu-bar/userscript.js:7]: Could this be moved above the loop?

File: addons/compact-menu-bar/addon.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "Compact editor menu bar",
```
Comment (DNin01) [inline addons/compact-menu-bar/addon.json:2]: If the only thing this addon does is hide the username, we might want to use a different name as making the whole menu bar compact sounds more involved than just hiding the username. In the Customizable navigation bar addon, that setting is called "compact user dropdown".

File: addons/compact-menu-bar/userscript.js (hunk 2 above)
Comment (iqnite) [reply addons/compact-menu-bar/userscript.js:7]: Done.

File: addons/compact-menu-bar/userscript.js (hunk 1 above)
Comment (iqnite) [reply addons/compact-menu-bar/userscript.js:10]: This didn't fix the problem. Apparently, Scratch regenerates the CSS every time you switch the page.
What I don't understand is why the label is shown properly on the website, but not in the editor.

File: addons/compact-menu-bar/addon.json (hunk 3 above)
Comment (iqnite) [reply addons/compact-menu-bar/addon.json:2]: Done.

**Comment by iqnite** [comment]
I tried using setInterval, but it is repeated only once! Can anybody help me?

File: addons/compact-menu-bar/userstyle.css [hunk 4]
```diff
@@ -0,0 +1,11 @@
+.account-nav_profile-name_2oRiV {
```
Comment (DNin01) [inline addons/compact-menu-bar/userstyle.css:1]: We usually use `[class*=...]` for these style rules. In this case:
```suggestion
[class*="account-nav_profile-name"] {
```
<sub>[Learn more](https://github.com/ScratchAddons/ScratchAddons/pull/5595)</sub>

**Comment by WorldLanguages** [comment]
> I tried using setInterval, but it is repeated only once! Can anybody help me?

You probably shouldn't be using setInterval anyway.

**Comment by Samq64** [comment]
Since it looks like this addon's functionality is the the same as the toggle for the custom navigation bar maybe they could be merged?

**Comment by mxmou** [comment]
> Since it looks like this addon's functionality is the the same as the toggle for the custom navigation bar maybe they could be merged?

Website and editor addons are usually separated.

**Comment by iqnite** [comment]
Everything is fixed now (thanks @CST1229 !)

**Review by DNin01 — CHANGES_REQUESTED** [review]
This addon should ideally support dynamic enable and disable. If you want to learn how to do that, check the [documentation](https://scratchaddons.com/docs/reference/addon-api/addon.self/#using-addonselfdisabled-and-events-dynamicdisable).

**Review by DNin01 — APPROVED** [review]
Looks good now!

**Comment by Samq64** [comment]
Related: #4018

**Comment by iqnite** [comment]
> I'm not sure if a new addon is the best place for this. Would merging this with `tutorials-button` be better?

That's what I had planned initially, and that's why the branch's called `custom-editor-bar`. I wanted to merge `editor-messages` and `tutorials-button` and add this new feature. However, I saw that it would be less complicated and save performance if I did this. I'll consider merging them as soon as the "Merge File & Edit menus" feature is finished (by the way I haven't even begun to make it so it's just an inspiration).

File: addons/compact-menu-bar/userscript.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+export default async function ({ addon }) {
+  if (await addon.auth.fetchIsLoggedIn()) {
+    const username = await addon.auth.fetchUsername();
+    addon.tab.redux.initialize(); // Start listening to Redux events
+    addon.tab.redux.addEventListener("statechanged", async (e) => {
+      if (addon.self.disabled) return; // Skip if the addon is disabled
+      if (e.detail.action.type !== "scratch-gui/menus/OPEN_MENU" || e.detail.action.menu !== "accountMenu") return; // Skip if another Redux event occurs
+      const profileSpans = await addon.tab.waitForElement(".menu_menu-item_3EwYA.menu_hoverable_3u9dt"); // Do this if the menu is opened
+      const profile = profileSpans.appendChild(document.createElement("span")); // Add username to "Profile" menu option
```
Comment (Samq64) [inline addons/compact-menu-bar/userscript.js:9]: I don't think the addon needs to support dynamic enable with the menu is open, but could `addon.tab.displayNoneWhileDisabled` be used here so the username style doesn't get messed up when dynamically disabling the addon with the menu open?

File: addons/compact-menu-bar/addon.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
+{
+  "name": "Compact menu bar user dropdown",
+  "description": "Hides the username from the account dropdown in the editor.",
+  "credits": [
+    {
+      "name": "philipp2007",
+      "link": "https://ggigabyte.repl.co/"
```
Comment (Samq64) [inline addons/compact-menu-bar/addon.json:7]: Who is philipp2007? Is that you?

File: addons/compact-menu-bar/userscript.js (hunk 5 above)
Comment (iqnite) [reply addons/compact-menu-bar/userscript.js:9]: I'll look into it.

Comment (iqnite) [reply addons/compact-menu-bar/userscript.js:9]: Tried it now, and it messes up the menu when toggled. I think it's good enough like this.

**Comment by Samq64** [comment]
Could this be closed? #6297 just extends this PR.

**Comment by iqnite** [comment]
> Could this be closed? #6297 just extends this PR.

No. If there's a bug in the new changes, this would still be mergeable. I'll close it as soon as the new one gets merged.

**Comment by iqnite** [comment]
Definitely moved to #6297.

**Pull request closed without merging**